File: binding/src/libqiconengineplugin.cpp
#include <QIconEngine>
#include <QIconEnginePlugin>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qiconengineplugin.h>
#include "libqiconengineplugin.h"
#include "libqiconengineplugin.hxx"

QIconEnginePlugin* QIconEnginePlugin_new() {
	 return new VirtualQIconEnginePlugin();
}

QIconEnginePlugin* QIconEnginePlugin_new2(QObject* parent) {
	 return new VirtualQIconEnginePlugin(parent);
}

libqt_string QIconEnginePlugin_Tr(const char* s) {
	QString _ret = QIconEnginePlugin::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QIconEngine* QIconEnginePlugin_Create(QIconEnginePlugin* self, const libqt_string filename) {
	return self->create(QString::fromUtf8(filename.data, filename.len));
}

libqt_string QIconEnginePlugin_Tr2(const char* s, const char* c) {
	QString _ret = QIconEnginePlugin::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QIconEnginePlugin_Tr3(const char* s, const char* c, int n) {
	QString _ret = QIconEnginePlugin::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
QIconEngine* QIconEnginePlugin_QBaseCreate(QIconEnginePlugin* self, const libqt_string filename) {
	auto* vqiconengineplugin = dynamic_cast<VirtualQIconEnginePlugin*>(self);
	if (vqiconengineplugin && vqiconengineplugin->isVirtualQIconEnginePlugin) {
vqiconengineplugin->setQIconEnginePlugin_Create_IsBase(true);
	return vqiconengineplugin->create(QString::fromUtf8(filename.data, filename.len));
}
}

// Auxiliary method to allow providing re-implementation
void QIconEnginePlugin_OnCreate(QIconEnginePlugin* self, intptr_t slot) {
	auto* vqiconengineplugin = dynamic_cast<VirtualQIconEnginePlugin*>(self);
	if (vqiconengineplugin && vqiconengineplugin->isVirtualQIconEnginePlugin) {
vqiconengineplugin->setQIconEnginePlugin_Create_Callback(reinterpret_cast<VirtualQIconEnginePlugin::QIconEnginePlugin_Create_Callback>(slot));
}
}

void QIconEnginePlugin_Delete(QIconEnginePlugin* self) {
    delete self;
}

File: binding/src/libqiconengineplugin.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qiconengineplugin.html
pub const qiconengineplugin = struct {

    /// New constructs a new QIconEnginePlugin object.
    pub fn New() QtC.QIconEnginePlugin {
        return qtc.QIconEnginePlugin_new();
    }


    /// New2 constructs a new QIconEnginePlugin object.
    pub fn New2(parent: ?*anyopaque) QtC.QIconEnginePlugin {
        return qtc.QIconEnginePlugin_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QIconEnginePlugin_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qiconengineplugin.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Create(self: ?*anyopaque, filename: []const u8) QtC.QIconEngine {
        const filename_str = qtc.libqt_string{
    .len = filename.len,
    .data = filename.ptr,
};
return qtc.QIconEnginePlugin_Create(@ptrCast(self), filename_str);
    }

    /// Allows for overriding the related default method
    pub fn OnCreate(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) QtC.QIconEngine) void {
        qtc.QIconEnginePlugin_OnCreate(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseCreate(self: ?*anyopaque, filename: []const u8) QtC.QIconEngine {
        const filename_str = qtc.libqt_string{
    .len = filename.len,
    .data = filename.ptr,
};
return qtc.QIconEnginePlugin_QBaseCreate(@ptrCast(self), filename_str);
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QIconEnginePlugin_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qiconengineplugin.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QIconEnginePlugin_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qiconengineplugin.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QIconEnginePlugin_Delete(@ptrCast(self));
    }
};
